### PR TITLE
Add support for date-time string format

### PR DESCRIPTION
**Pull request by alkallio**

### CODE REVIEW

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    public function validateFailureProvider()
+    {
+        //input
+        return [
+            ['2019-11-14T14:30:26Z+02:00', 'pattern'],
+            ['2019-11-14T', 'pattern'],
+            ['2019-11-14 14:30:26+02:00', 'pattern'],
+            ['2019-11-14T14:30:26+02:00random', 'pattern'],
+            ['2019-11-14T14:30:26.123543333+02:00', 'maxLength'],
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: Would be nice to check if returns only valid dates

`2019-02-30T01:01:01Z`

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ */
+class StringWithDatetimeFormatValidatorTest extends TestCase
+{
+
+    /**
+     * @var String
+     */
+    protected $object;
+
+    protected function setUp()
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: : void

File: src/Formats/FormatValidator.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                //Todo Create a TDO for formatProperties
+                $dateTimeFormatValidator = (new DateTime())
+                    ->validateFormat(
+                        $data,
+                        $formatProperties
+                    );
+
+                if (!$dateTimeFormatValidator->isValid()) {
+                    $this->isValid = false;
+                    $this->failure = $dateTimeFormatValidator->getFailure();
```
Comment (nohponex) [inline src/Formats/FormatValidator.php]: ->failure = 'format' maybe?

File: src/Formats/FormatValidator.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace Phramework\Validate\Formats;
+
+use Phramework\Validate\IFormat;
+
+class FormatValidator implements IFormat
```
Comment (nohponex) [inline src/Formats/FormatValidator.php]: `date-time`, `date` etc are formats only for string

so could it be `StringFormatValidator` instead?

File: src/Formats/DateTime.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        if ($formatProperties->formatMaximum !== null) {
+            $formatMaximum = (new \DateTime(
+                $formatProperties->formatMaximum
+            ))->getTimestamp();
+
+            if ($date > $formatMaximum) {
+                $this->isValid = false;
+                $this->failure = 'Higher than maximum';
```
Comment (nohponex) [inline src/Formats/DateTime.php]: = `formatMaximum`

Comment (nohponex) [reply src/Formats/DateTime.php]: we have the schema property name as failure

File: src/Formats/DateTime.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        $date = (new \DateTime($data))->getTimestamp();
+
+        if ($formatProperties->formatMinimum !== null) {
+            $formatMinimum = (new \DateTime(
+                $formatProperties->formatMinimum
+            ))->getTimestamp();
+
+            if ($date < $formatMinimum) {
+                $this->isValid = false;
```
Comment (nohponex) [inline src/Formats/DateTime.php]: why this class isn't stateless?

File: src/Formats/FormatValidator.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        return $this->failure;
+    }
+
+    public function validateFormat(
+        string $data,
+        string $format,
+        \stdClass $formatProperties = null
+    ): IFormat {
+        switch ($format) {
+            case "date-time":
```
Comment (nohponex) [inline src/Formats/FormatValidator.php]: single quotes

File: src/Formats/DateTime.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        try {
+            $date = (new \DateTime($data))->getTimestamp();
+        } catch (\Exception $e) {
+            return new ValidateResult(
+                $data,
+                false,
+                new IncorrectParametersException(
+                    [
+                        'type' => $type,
+                        'failure' => 'Invalid date',
```
Comment (nohponex) [inline src/Formats/DateTime.php]: date-time

File: src/Formats/DateTime.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Phramework\Validate\Formats;
+
+use Phramework\Exceptions\IncorrectParametersException;
+use Phramework\Validate\ValidateResult;
+
+class DateTime
```
Comment (nohponex) [inline src/Formats/DateTime.php:15]: @internal

File: src/Formats/StringFormatValidator.php [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        string $format,
+        \stdClass $formatProperties = null
+    ): ValidateResult {
+        $validateResult = new ValidateResult(
+            $data,
+            true
+        );
+
+        switch ($format) {
+            case "date-time":
```
Comment (nohponex) [inline src/Formats/StringFormatValidator.php]: single quote

File: src/IFormat.php [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace Phramework\Validate;
+
+interface IFormat
```
Comment (nohponex) [inline src/IFormat.php]: IFormatValidator

**Review by nohponex — CHANGES_REQUESTED** [review]

File: src/Formats/FormatValidator.php (hunk 7 above)
Comment (alkallio) [reply src/Formats/FormatValidator.php]: Fixed

File: src/Formats/DateTime.php (hunk 8 above)
Comment (alkallio) [reply src/Formats/DateTime.php]: Updated

File: src/Formats/FormatValidator.php (hunk 4 above)
Comment (alkallio) [reply src/Formats/FormatValidator.php]: Updated name to better reflect purpose

File: src/Formats/DateTime.php (hunk 5 above)
Comment (alkallio) [reply src/Formats/DateTime.php]: Fixed

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+namespace Phramework\Validate;
+
+use PHPUnit\Framework\TestCase;
+
+class StringWithDatetimeFormatValidatorTest extends TestCase
+{
+
+    /**
+     * @var String
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: ?

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            null,
+            false,
+            'date-time'
+        );
+
+        $this->object->setFormatMinimum('2018-11-14T14:30:26+02:00');
+        $this->object->setFormatMaximum('2020-11-14T14:30:26+02:00');
+    }
+
+    public function validateSuccessProvider()
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: : array

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+            ['2019-11-14T00:80:00-00:00', 'date-time'],
+            ['2017-11-14T14:30:26+02:00', 'formatMinimum'],
+            ['2023-11-14T14:30:26+02:00', 'formatMaximum'],
+        ];
+    }
+
+    /**
+     * @dataProvider validateSuccessProvider
+     */
+    public function testValidateSuccess($input, $expected)
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: : void

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+        $this->assertInternalType('string', $return->value);
+        $this->assertEquals($expected, $return->value);
+        $this->assertTrue($return->status);
+    }
+
+    /**
+     * @dataProvider validateFailureProvider
+     */
+    public function testValidateFailure($input, $failure)
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php]: strong type

**Review by nohponex — CHANGES_REQUESTED** [review]

File: tests/src/StringWithDatetimeFormatValidatorTest.php (hunk 12 above)
Comment (alkallio) [reply tests/src/StringWithDatetimeFormatValidatorTest.php]: Fixed

File: tests/src/StringWithDatetimeFormatValidatorTest.php (hunk 14 above)
Comment (alkallio) [reply tests/src/StringWithDatetimeFormatValidatorTest.php]: Added strong types in every function in the new test

**Review by nohponex — APPROVED** [review]

File: tests/src/StringWithDatetimeFormatValidatorTest.php [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        return [
+            ['2019-11-14T14:30:26+02:00', '2019-11-14T14:30:26+02:00'],
+            ['2019-11-14T14:30:60+02:00', '2019-11-14T14:30:60+02:00'],
+            ['2019-11-14T14:30:60+00:00', '2019-11-14T14:30:60+00:00'],
+            ['2019-11-14T14:30:45-05:00', '2019-11-14T14:30:45-05:00'],
+            ['2019-11-14T00:00:00-05:00', '2019-11-14T00:00:00-05:00'],
+            ['2019-11-14T14:30:26Z', '2019-11-14T14:30:26Z'],
+            ['2018-11-14T14:30:26+02:00', '2018-11-14T14:30:26+02:00'],
+            ['2020-11-14T14:30:26+02:00', '2020-11-14T14:30:26+02:00'],
+            ['2020-02-28T00:30:26+02:00', '2020-02-28T00:30:26+02:00'],
```
Comment (nohponex) [inline tests/src/StringWithDatetimeFormatValidatorTest.php:41]: You didn't include `2006-01-02T15:04:05+0500`
You didn't include `2006-01-02T15:04:05+05`

**Pull request merged**